Translate along and rotate around any arbitrary vector

smortvedt opened the pull request:
This PR fixes #895 by adding the following functions:
- Rotation using an arbitrary vector as both the axis and center of rotation.
- Rotation using an arbitrary vector as the axis but using the object's center as the center of rotation.
- Rotation as above but using an arbitrary plane's normal as the vector.
- Translation along an arbitrary plane in the plane's x, y, and z coordinates.
- Translation along an arbitrary vector using the vector as the z direction.

The Rotation/Translation UI was modified to have dropdown menus to select the plane/line references if they exist in the db root. Only single segment polylines are allowed to be selected. There is a button to toggle between the advanced mode and the regular mode as well. Any comments or suggestions would be appreciated! Thanks for your time to review.

dgirardeau commented on qCC/ccGraphicalTransformationTool.cpp:
```diff
@@ -225,12 +225,91 @@ void ccGraphicalTransformationTool::stop(bool state)
+	bool arbitraryVectorTranslation = true;
+	if (arbitraryVectorTranslation) 
+	{
+		CCVector3d vec(-.20, -.20, -.20);
```
Is there a rationale behind this -0.2 quantity? Or is it really arbitrary?

smortvedt replied:
I believe you're looking at outdated code, this was test code from my first commit. Before I updated the UI to import a vector I manually changed the test vector.

dgirardeau commented on qCC/ccGraphicalTransformationTool.cpp:
```diff
@@ -210,6 +226,68 @@ bool ccGraphicalTransformationTool::setAdvancedTranslationTransform(ccHObject* t
+			CCVector3d newCenter = CCVector3d::fromArray(m_toTransform.getBB_recursive().getCenter().u);
+			setRotationCenter(newCenter);
+		}
+		rotComboBox->setCurrentIndex(3); //Z
```
Maybe use an enum here to be clearer? (and at the similar places as well)

smortvedt replied:
Sure, I can add an enum throughout

dgirardeau commented on qCC/ccGraphicalTransformationTool.cpp:
```diff
@@ -218,7 +296,7 @@ void ccGraphicalTransformationTool::advTranslateRefChanged(int index)
 	int id = translateComboBox->itemData(index).toInt();
 	int selectedObjectIndex = -1;
-	for (int i; i < m_planesAndLineSegments.size(); i++)
+	for (int i = 0; i < m_planesAndLineSegments.size(); i++)
```
I believe it should be size_t instead of int (if m_planesAndLineSegments is a vector, or a ccHObject::Container = vector)

dgirardeau replied:
Same comment for other places

smortvedt replied:
Sure, ill change that

dgirardeau commented:
Ok, I tested on my side.

First thing: I believe you should increase the 'overlay UI' width, otherwise it's hard to read the options and the entities names:
![image](https://user-images.githubusercontent.com/2735509/74019834-93b21680-4998-11ea-9b4e-0d16f770a702.png)
(weel, that's probably because I have a 4K screen now :( )

dgirardeau commented:
Second:
Depending on the screen rotation, the translation along the plane was a bit weird (inversion of up/down compared to left/right --> I mean the translation along left/right was following the mouse motion, but the translation along up/down direction was inverted).

And if you change the virtual camera position (press the space bar of click on the "pause" button, move the camera, and then resume), the translation doesn't seem to stick to the plane directions anymore.

smortvedt commented:
Sorry about the translate weirdness, I think you were seeing the mouse moves still in reference to the origin frame and not the plane frame. I added a transform to the mouse moves so they should match up with the plane reference frame if using that option. 

Here's what the UI looks like on my screen at width 200:
![image](https://user-images.githubusercontent.com/36835723/74171961-42966280-4be4-11ea-8de2-a3c1fb5bccd5.png)

I bumped it up to 250, it looks like maybe that would do it for yours as well, but I am kinda just guessing, is there a better way to size the width? Percentage maybe? Here's what my screen looks like a 250 

![image](https://user-images.githubusercontent.com/36835723/74173523-18926f80-4be7-11ea-8b47-9d1402d6545f.png)

![image](https://user-images.githubusercontent.com/36835723/74173554-247e3180-4be7-11ea-9bc1-753f51f3ce3f.png)

![image](https://user-images.githubusercontent.com/36835723/74173599-3f50a600-4be7-11ea-9cef-0010a359dfe0.png)

dgirardeau commented:
I'll try to update the dialog size so that it works better on my side (there seems to be something else happening, as the radio buttons are still displayed in a weird way even though the dialog is much larger):
![image](https://user-images.githubusercontent.com/2735509/74195691-19b1b400-4c5c-11ea-91ad-e6a44652d1d9.png)

dgirardeau commented:
And the 'translate along' option doesn't seem to work anymore on my side!

WargodHernandez commented:
@smortvedt 
I believe you are missing a QHBoxLayout around the radio buttons, I haven't checked it but that is a common side effect of not having that

smortvedt commented:
Just updated the layout. The groupbox widget was causing the layout spacing to be off, so I took that out and added a slot to effectively group the two radios.

As for the translation not working on your side, is the object not moving at all now? Or is it translating but still along the origin frame? I can add print statements to help track down the issue, on my side its working though :/

smortvedt commented:
@WargodHernandez just found a bug, when the reference object is a child of the object being rotated, the reference object is rotated as well and the resulting translations afterwards are not as expected. I'll be working on this bug, maybe this is what you are seeing as well!

dgirardeau commented:
On my side it is more that I can freely translate the object 'as usual', and the constraint of the plane doesn't seem to be taken into account (the planes were not children of the cloud)

dgirardeau commented:
I can try to debug it on my side but I'll need some time.

smortvedt commented:
Just committed the bug fix, I tested the all combinations of use that I could think of (rotation plane as child, rotation line as non-child, translation line as child, etc. etc.) as well as the different operations in series and it all passed my checks. 
The tests I ran boil down to rotating and translating along planes/lines when the are:
- a child of the selected object(s) (moving with the object)
- selected with the object(s) to translate/rotate (moving with the object)
- separate from the object(s) to translate/rotate (not moving with the object)

Chris is going to pull the code and test it later today. If its still not working for you I can add print statements in spots that may highlight what's the issue is.

dgirardeau commented:
Maybe I'm missing something.

I use this test dataset: 
[maquette_with_planes.zip](https://github.com/CloudCompare/CloudCompare/files/4201021/maquette_with_planes.zip)

I select the cloud, start the 'Interactive Transformation' tool, and then select the 'Advanced' mode and select one of the plane in the 'Translate along' field:
![image](https://user-images.githubusercontent.com/2735509/74477925-920eb400-4eac-11ea-9ef7-760aa71c1261.png)

I was expecting to be constrained along this plane when translating the cloud. But it's as if nothing changed and as if the tool is just behaving 'normally'.

WargodHernandez commented:
> I was expecting to be constrained along this plane when translating the cloud. But it's as if nothing changed and as if the tool is just behaving 'normally'.

it appears you have all the translation active 
![image](https://user-images.githubusercontent.com/7495752/74480182-4e448c00-4e65-11ea-83b5-5f1e428ff35a.png)
That would make it similar to the default, freedom on all axis.
if you restrict your selection to the x &/or y axis then the cloud will slide along the plane
if you select z you will slide along its normal.

or did you try that and I missed something?

dgirardeau commented:
Ah! Sorry, I didn't understood this option was relying on the above Tx, Ty and Tz checkboxes. (maybe because the first time I tested the tool it was already doing something even in these checkboxes were all checked?).

Anyway, now I get it,and it works ;)

dgirardeau commented:
Thanks for the contribution again!
